Occasional crash in `WebViewImpl::setMarkedText` (`com.apple.WebKit: _ZN6WebKit11WebViewImpl13setMarkedTextEP11objc_object8_NSRangeS3_`)

rr-codes opened the pull request:
#### b33075ac6c0e3bc0177339797b1ae644ec8faf2f
<pre>
Occasional crash in `WebViewImpl::setMarkedText` (`com.apple.WebKit: _ZN6WebKit11WebViewImpl13setMarkedTextEP11objc_object8_NSRangeS3_`)
<a href="https://bugs.webkit.org/show_bug.cgi?id=274186">https://bugs.webkit.org/show_bug.cgi?id=274186</a>
<a href="https://rdar.apple.com/127658062">rdar://127658062</a>

Reviewed by Abrar Rahman Protyasha.

The existing logic to ensure that `setWritingSuggestions` only gets called when `string` is an
`NSAttributedString` was flawed in some cases (when `m_isHandlingAcceptedCandidate == true` and `hasTextCompletion == false`).
This resulted in `-[NSAttributedString string]` being called on an `NSString`, causing a crash.

* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::setMarkedText):
* Tools/TestWebKitAPI/Tests/mac/WKWebViewMacEditingTests.mm:
(-[SetMarkedTextWithNoAttributedStringTestCandidate initWithReplacementString:inRange:]):
(-[SetMarkedTextWithNoAttributedStringTestCandidate replacementString]):
(-[SetMarkedTextWithNoAttributedStringTestCandidate resultType]):
(-[SetMarkedTextWithNoAttributedStringTestCandidate range]):
(TEST(WKWebViewMacEditingTests, SetMarkedTextWithNoAttributedString)):

Canonical link: <a href="https://commits.webkit.org/278792@main">https://commits.webkit.org/278792@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/45b627c862e513c940abb10c9d10b62ba6b6fd07

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/51587 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/30897 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/3940 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/54854 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/2280 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/53890 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/37234 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/1961 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/41992 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/53686 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/28533 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/44496 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/23120 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/25827 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/1757 "Passed tests") | [⏳ 🛠 wpe-skia ](https://ews-build.webkit.org/#/builders/WPE-Skia-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/47794 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/1847 "Found 1 new test failure: media/video-webkit-playsinline.html (failure)") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/56446 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/26707 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/1714 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/49390 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/27946 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/44564 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/48588 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/11285 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/28840 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/7523 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/27682 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Tools/TestWebKitAPI/Tests/mac/WKWebViewMacEditingTests.mm:
```diff
@@ -219,6 +219,70 @@ - (SlowTextInputContext *)_web_superInputContext
+@end
+
+@implementation SetMarkedTextWithNoAttributedStringTestCandidate {
+    NSString *_string;
```
I know this is testing code, but it's a good habit to not introduce new raw pointer usage.

rr-codes replied:
lol all of this is just copy-pasted from another test -- will fix though

pxlcoder commented on Tools/TestWebKitAPI/Tests/mac/WKWebViewMacEditingTests.mm:
```diff
@@ -219,6 +219,70 @@ - (SlowTextInputContext *)_web_superInputContext
+
+    [webView stringByEvaluatingJavaScript:modifySelectionJavascript];
+
+    [webView _handleAcceptedCandidate:[[SetMarkedTextWithNoAttributedStringTestCandidate alloc] initWithReplacementString:@"test" inRange:NSMakeRange(0, 0)]];
```
And this is a memory leak.

rr-codes replied:
(again I blame this on the copy/paste 😂 )

pxlcoder commented on Source/WebKit/UIProcess/mac/WebViewImpl.mm:
```diff
@@ -5224,27 +5224,27 @@ static BOOL shouldUseHighlightsForMarkedText(NSAttributedString *string)
-        if (!isAttributedString)
-            return NO;
+    if (isAttributedString) {
+        NSAttributedString *attributedString = (NSAttributedString *)string;
```
Could be rewritten to have variable assignment inside the if. And use `dynamic_objc_cast`.

rr-codes replied:
wdym?

whsieh replied:
```
if (RetainPtr attributedString = dynamic_objc_cast<NSAttributedString>(string)) {
    …
}
```

rr-codes replied:
isn't that technically resulting in an extra call to `isKindOfClass`? (since the existing code still needs the `BOOL isAttributedString = [string isKindOfClass:[NSAttributedString class]];` to exist. I guess it doesn't really matter though

rr-codes committed:
Occasional crash in `WebViewImpl::setMarkedText` (`com.apple.WebKit: _ZN6WebKit11WebViewImpl13setMarkedTextEP11objc_object8_NSRangeS3_`)
https://bugs.webkit.org/show_bug.cgi?id=274186
rdar://127658062

Reviewed by Abrar Rahman Protyasha.

The existing logic to ensure that `setWritingSuggestions` only gets called when `string` is an
`NSAttributedString` was flawed in some cases (when `m_isHandlingAcceptedCandidate == true` and `hasTextCompletion == false`).
This resulted in `-[NSAttributedString string]` being called on an `NSString`, causing a crash.

* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::setMarkedText):
* Tools/TestWebKitAPI/Tests/mac/WKWebViewMacEditingTests.mm:
(-[SetMarkedTextWithNoAttributedStringTestCandidate initWithReplacementString:inRange:]):
(-[SetMarkedTextWithNoAttributedStringTestCandidate replacementString]):
(-[SetMarkedTextWithNoAttributedStringTestCandidate resultType]):
(-[SetMarkedTextWithNoAttributedStringTestCandidate range]):
(TEST(WKWebViewMacEditingTests, SetMarkedTextWithNoAttributedString)):

Canonical link: https://commits.webkit.org/278792@main
